pkg/distributor/distributor.go

```diff
@@ -367,12 +371,14 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
 	keys := make([]uint32, 0, len(req.Timeseries))
 	validatedSamples := 0
 	for _, ts := range req.Timeseries {
-		key, err := d.tokenForLabels(userID, ts.Labels)
-		if err != nil {
+		key, validatedSeries, err := d.validateSeries(ts, userID, removeReplica)
+
+		// In case of a fatal error (ie. logic bug) we should fail the entire
+		// request and not consider it a partial error.
+		if err == errShardLabelsNotSorted {
```

Comment:
This seem like it might be a little brittle to me - previously all errors from `validateSeries` were "soft", now some are special.

Reply:
Right. From a design perspective, I think the main issue is that both labels removal and sharding token generation is done inside the `validateSeries()` which sounds "weird" (it does more than "validation").

I've pushed a commit where both labels removal and sharding token generation has been moved outside.

What's your take now?